crates/ol/mempool/src/state.rs

```diff
@@ -0,0 +1,1933 @@
+        for txid in self.ordering_index.values() {
+            if let Some((_, tx)) = ready_txs.iter().find(|(tid, _)| tid == txid) {
+                result.push((*txid, tx.clone()));
+            }
+        }
+
+        Ok(result)
+    }
+
+    /// Remove transactions from the mempool.
+    ///
+    /// Returns the removed transaction IDs.
+    ///
+    /// Also removes from the per-account index.
+    pub(crate) fn remove_transactions(
+        &mut self,
+        ctx: &MempoolContext,
+        ids: &[OLTxId],
+    ) -> OLMempoolResult<Vec<OLTxId>> {
+        let mut removed = Vec::with_capacity(ids.len());
+        let mut account_min_seq: HashMap<AccountId, u64> = HashMap::new();
+
+        // First pass: Remove specified transactions and collect minimum seq_no per account
+        for txid in ids {
+            // Check if entry exists and get necessary data before removing
+            if let Some(entry) = self.entries.get(txid) {
+                // Remove from database FIRST (ensures durability - if this fails, memory unchanged)
+                ctx.storage.mempool().del_tx(*txid)?;
+
+                // Recompute priority for ordering index removal
+                let priority = ctx.ordering_strategy.compute_priority(entry);
+                let size_bytes = entry.size_bytes;
+                let account_id = entry.tx.target();
+
+                // Track minimum seq_no for each affected account (for cascade removal)
+                let account_info = if let Some(base_update) = entry.tx.base_update() {
+                    let seq_no = base_update.operation().seq_no();
+                    Some((account_id, seq_no))
+                } else {
+                    None
+                };
+
+                // Now remove from memory (safe because DB operation succeeded)
+                self.entries.remove(txid);
+                self.ordering_index.remove(&priority);
+
+                // Update account tracking
+                if let Some((account, seq_no)) = account_info {
+                    account_min_seq
+                        .entry(account)
+                        .and_modify(|min_seq| *min_seq = (*min_seq).min(seq_no))
+                        .or_insert(seq_no);
+                }
+
+                // Remove from per-account index
+                if let Some(account_txs) = self.by_account.get_mut(&account_id) {
+                    account_txs.remove(txid);
+                    if account_txs.is_empty() {
+                        self.by_account.remove(&account_id);
+                    }
+                }
+
+                // Update stats
+                self.stats.mempool_size -= 1;
+                self.stats.total_bytes -= size_bytes;
+
+                removed.push(*txid);
+            }
+        }
+
+        // Second pass: Cascade-remove dependent transactions and recalculate pending_seq_no
+        for (account, min_failed_seq) in account_min_seq {
```

Comment:
>for (account, min_failed_seq) in account_min_seq {

What happens if there is no failure? It can be that we processed txs (with seq no) 2,3 but we haven't processed 4,5. In this case we don't want to remove 4 and 5.